### PR TITLE
User-level notification settings

### PR DESCRIPTION
**Pull request by ViacheslavKlimov**
## Pull Request description

image

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [x] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [x] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [x] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: dao/src/main/java/org/thingsboard/server/dao/notification/DefaultNotificationSettingsService.java [hunk 1]
```diff
@@ -81,6 +95,50 @@ public NotificationSettings findNotificationSettings(TenantId tenantId) {
...
 
+    @Override
+    public UserNotificationSettings saveUserNotificationSettings(TenantId tenantId, UserId userId, UserNotificationSettings settings) {
+        User user = userService.findUserById(tenantId, userId);
+        ObjectNode additionalInfo = (ObjectNode) Optional.ofNullable(user.getAdditionalInfo()).orElseGet(JacksonUtil::newObjectNode);
+        additionalInfo.set(USER_SETTINGS_KEY, JacksonUtil.valueToTree(settings));
+        user.setAdditionalInfo(additionalInfo);
+        userService.saveUser(user);
+        return formatUserNotificationSettings(settings);
+    }
```
Comment (ViacheslavKlimov) [inline dao/src/main/java/org/thingsboard/server/dao/notification/DefaultNotificationSettingsService.java:106]: Settings are stored inside additionalInfo to minimize the amount of additional DB calls.

File: application/src/test/resources/application-test.properties [hunk 2]
```diff
@@ -67,3 +67,5 @@ sql.ttl.audit_logs.ttl=2592000
 
 sql.edge_events.partition_size=168
 sql.ttl.edge_events.edge_event_ttl=2592000
+
+server.log_controller_error_stack_trace=false
```
Comment (ViacheslavKlimov) [inline application/src/test/resources/application-test.properties:71]: Added to be able to quickly enable the value when investigating test failures.

**Review by vvlladd28 — APPROVED** [review]

File: dao/src/main/java/org/thingsboard/server/dao/notification/DefaultNotificationSettingsService.java (hunk 1 above)
Comment (ashvayka) [reply dao/src/main/java/org/thingsboard/server/dao/notification/DefaultNotificationSettingsService.java:104]: See ```org.thingsboard.server.common.data.settings.UserSettings```. We have them cached. Please add a new ```UserSettingsType``` and consider using them.

File: dao/src/main/java/org/thingsboard/server/dao/notification/DefaultNotificationSettingsService.java [hunk 3]
```diff
@@ -81,6 +91,53 @@ public NotificationSettings findNotificationSettings(TenantId tenantId) {
...
+        userSettingsService.saveUserSettings(tenantId, userSettings);
+        return formatUserNotificationSettings(settings);
+    }
+
+    @Override
+    public UserNotificationSettings getUserNotificationSettings(TenantId tenantId, UserId userId, boolean format) {
+        UserSettings userSettings = userSettingsService.findUserSettings(tenantId, userId, UserSettingsType.NOTIFICATIONS);
+        UserNotificationSettings settings;
+        if (userSettings != null) {
+            settings = JacksonUtil.treeToValue(userSettings.getSettings(), UserNotificationSettings.class);
```
Comment (ashvayka) [inline dao/src/main/java/org/thingsboard/server/dao/notification/DefaultNotificationSettingsService.java:109]: What if we change the structure of UserNotificationSettings? I would recommend adding try catch and fallback to the UserNotificationSettings.DEFAULT.

File: common/data/src/main/java/org/thingsboard/server/common/data/page/SortOrder.java [hunk 4]
```diff
@@ -36,4 +36,6 @@ public static enum Direction {
         ASC, DESC
     }
 
+    public static final SortOrder byCreatedTimeDesc = new SortOrder("createdTime", Direction.DESC);
```
[the remark this answers is not in the document]
Comment (ViacheslavKlimov) [reply common/data/src/main/java/org/thingsboard/server/common/data/page/SortOrder.java:39]: I am familiar with the convention, it just looks better this way. Besides, we have multiple places in the project where we use such naming (in camel case). But ok, will fix this up

**Pull request merged**